man/dhcpd-pools.1
.\"	Sami Kerola
.\"	http://www.iki.fi/kerolasa/
.\"	kerolasa@iki.fi
.\"
.\"	Add'l ontributions by:
.\"		Dan Thorson
.\"
.TH DHCPD-POOLS "1" "October 2012" "dhcpd-pools" "User Commands"
.SH NAME
dhcpd-pools \- ISC dhcpd pools usage analysis
.SH SYNOPSIS
.B dhcpd-pools
[options]
.SH DESCRIPTION
The program analyses ISC dhcpd shared network and pool usage and outputs the
results in a format selected by user.
.SH OUTPUT FIELDS
.TP
.I "shared net name"
Name of the shared-network for the range.
.TP
.I "first ip"
First IP in lease pool range.
.TP
.I "last ip"
Last IP in lease pool range.
.TP
.I "max"
Number of IPs which exist in a pool, shared network or all together.
.TP
.I "cur"
Number of leases currently in use.
.TP
.I "percent"
Percent of IPs currently in use compared to max.
.TP
.I "touch"
Number of IP's which appear in the lease file, but who's leases have expired.
A touched IP is either expired or abandoned.  The touched IP count is
somewhat misleading when you try to determine if an IP pool is big enough; it
is a better indicator of whether a pool is too large.
.TP
.I "t+c"
The sum of Touched and Currently in-use leases.
.TP
.I "t+c perc"
Percent of IPs either touched or currently in use, compared to max.
.TP
.I "bu"
Failover pair can allocate these addresses.  The count appears only if there
is failover configuration.
.TP
.I "bu perc"
Percent of addresses that failover pair can allocate.  The percent appears
only if there is failover configuration.
.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-config\fR=\fIFILE\fR
Path to the dhcpd.conf file.
.TP
\fB\-l\fR, \fB\-\-leases\fR=\fIFILE\fR
Path to the dhcpd.leases file.
.TP
\fB\-s\fR, \fB\-\-sort\fR=\fI[nimcptTe]\fR
Sort ranges by chosen fields as a sorting keys.  Maximum of five sort keys
can be defined.  Keys weight from left to right i.e. if more weighting keys
are equal next one is used.  The IP field is default sort key.
.TP
\fB\-r\fR, \fB\-\-reverse\fR
Sort results in reverse order.
.TP
\fB\-f\fR, \fB\-\-format\fR=\fI[thHcxX]\fR
Output format.  The default is
text
.RI ( t ).
Standard html
.RI ( h )
outputs only the HTML tables, and is useful for embeding more complex web
pages.  Full-html
.RI ( H )
provides complete HTML headers, etc., including in-line CSS.  The
.RI ( c )
stands for comma-separated values.  Output format xml
.RI ( x )
is similar to the dhcpstatus Perl module output.  The extended xml
.RI ( X )
format will print ethernet address details.
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIFILE\fR
.I File
where output is written.  Default is stdout.
.TP
\fB\-L\fR, \fB\-\-limit\fR=\fINR\fR
The
.I NR
will limit what will be printed.  Syntax is similar to
.IR chmod (1)
permission string.  The
.I NR
limit string uses two digits which vary between
.IR 0 \ to \ 7 .
The first digit determines which headers to display, and the second digit
determines which numeric analysis tables to include in the output.  The
following values are "OR'd" together to create the desired output.  The
default is
.IR 77 .
.PP
.RS
.PD 0
.TP
.B 01
Print ranges
.TP
.B 02
Print shared networks
.TP
.B 04
Print total summary
.TP
.B 10
Print range header
.TP
.B 20
Print shared network header
.TP
.B 40
Print total summary header
.PD
.RE
.TP
\fB\-\-warning\fR=\fIpercent\fR
Turn on alarm output format, and specify percentage number which will
cause an alarm.  If either a range or shared network will exceed
warning level return value of the command is
.BR 1 .
If only range monitoring is needed one can limit output by using option
.IR \-L10 ,
and for shared network monitoring similar is
.IR \-L20 .
If critical level is not speficied it defaults to
.BR 90 .
The
.I percent
argument allows fractions to be used.  Alarm output is intended to be
used with monitoring, such as Nagios.
.TP
\fB\-\-critical\fR=\fIpercent\fR
The option is similar to warning, with exception of return value which
is
.BR 2 .
If warning level is not specified it defaults to
.BR 80 .
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version information to standard output and exit successfully.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help to standard output and exit successfully.
.SH EXAMPLES
.TP
Print ranges header, and analysis.
$ dhcpd-pools -L 11 -c dhcpd.conf -l dhcpd.leases
.br
Ranges:
.br
shared net name [...]
.TP
Print shared networks and totals, both headers and results
$ dhcpd-pools -L 66 -c dhcpd.conf -l dhcpd.leases shared net name
.br
[...]
.TP
Alarming
$ dhcpd-pools -c dhcpd.conf -l dhcpd.leases --critical 80.1 --warning 75
.br
CRITICAL: dhcpd-pools: Ranges; crit: 14 warn: 22 ok: 220 Shared nets; crit: 1 warn: 0 ok: 4
.SH FILES
.TP
/etc/dhcpd.conf
ISC dhcpd configuration file.
.TP
/var/lib/dhcp/dhcpd.leases
ISC dhcpd lease file.
.SH AUTHOR
Written by Sami Kerola.
.br
XML support by Dominic Germain, Sogetel inc.
.PP
The software has FreeBSD License.
.SH "REPORTING BUGS"
Report bugs to
.MT kerolasa@iki.fi
Sami Kerola
.ME
.br
.UR http://dhcpd-pools.sourceforge.net/
Home page
.UE
.SH "SEE ALSO"
.BR dhcpd.leases (5),
.BR dhcpd.conf (5),
.BR chmod (1)


src/dhcpd-pools.c
/*
 * The dhcpd-pools has BSD 2-clause license which also known as "Simplified
 * BSD License" or "FreeBSD License".
 *
 * Copyright 2006- Sami Kerola. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the
 *       distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR AND CONTRIBUTORS OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing
 * official policies, either expressed or implied, of Sami Kerola.
 */

#include <config.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>
#include <errno.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <limits.h>

#include "close-stream.h"
#include "closeout.h"
#include "defaults.h"
#include "dhcpd-pools.h"
#include "xalloc.h"

int main(int argc, char **argv)
{
	int i, c, sorts = 0;
	int option_index = 0;
	char const *tmp;
	struct range_t *tmp_ranges;
	enum {
		OPT_WARN = CHAR_MAX + 1,
		OPT_CRIT
	};
	int ret_val;

	/* Options for getopt_long */
	static struct option const long_options[] = {
		{"config", required_argument, NULL, 'c'},
		{"leases", required_argument, NULL, 'l'},
		{"format", required_argument, NULL, 'f'},
		{"sort", required_argument, NULL, 's'},
		{"reverse", no_argument, NULL, 'r'},
		{"output", required_argument, NULL, 'o'},
		{"limit", required_argument, NULL, 'L'},
		{"version", no_argument, NULL, 'v'},
		{"help", no_argument, NULL, 'h'},
		{"warning", required_argument, NULL, OPT_WARN},
		{"critical", required_argument, NULL, OPT_CRIT},
		{NULL, 0, NULL, 0}
	};

	atexit(close_stdout);

	/* FIXME: make these allocations dynamic up on need. */
	config.dhcpdconf_file = xmalloc(sizeof(char) * MAXLEN);
	config.dhcpdlease_file = xmalloc(sizeof(char) * MAXLEN);
	config.output_file = xmalloc(sizeof(char) * MAXLEN);

	/* Make sure string has zero length if there is no
	 * command line option */
	config.output_file[0] = '\0';
	/* Alarming defaults. */
	config.warning = ALARM_WARN;
	config.critical = ALARM_CRIT;

	/* File location defaults */
	strncpy(config.dhcpdconf_file, DHCPDCONF_FILE, MAXLEN - 1);
	strncpy(config.dhcpdlease_file, DHCPDLEASE_FILE, MAXLEN - 1);
	tmp = OUTPUT_LIMIT;
	config.output_limit[0] = (*tmp - '0');
	tmp++;
	config.output_limit[1] = (*tmp - '0');
	fullhtml = false;

	/* Make sure some output format is selected by default */
	strncpy(config.output_format, OUTPUT_FORMAT, (size_t) 1);

	/* Default sort order is by IPs small to big */
	config.reverse_order = false;

	/* Parse command line options */
	while (1) {
		c = getopt_long(argc, argv, "c:l:f:o:s:rL:vh",
				long_options, &option_index);

		if (c == EOF)
			break;

		switch (c) {
		case 'c':
			/* config file */
			strncpy(config.dhcpdconf_file, optarg, MAXLEN - 1);
			break;
		case 'l':
			/* lease file */
			strncpy(config.dhcpdlease_file, optarg, MAXLEN - 1);
			break;
		case 'f':
			/* Output format */
			strncpy(config.output_format, optarg, (size_t) 1);
			break;
		case 's':
			/* Output sorting option */
			sorts = strlen(optarg);
			if (5 < sorts) {
				warnx
				    ("main: only first 5 sort orders will be used");
				strncpy(config.sort, optarg, (size_t) 5);
				sorts = 5;
			} else {
				strncpy(config.sort, optarg, (size_t) sorts);
			}
			for (i = 0; i < sorts; i++) {
				field_selector(config.sort[i]);
			}
			break;
		case 'r':
			/* What ever sort in reverse order */
			config.reverse_order = true;
			break;
		case 'o':
			/* Output file */
			strncpy(config.output_file, optarg, MAXLEN - 1);
			break;
		case 'L':
			/* Specification what will be printed */
			for (i = 0; i < 2; i++) {
				if (optarg[i] >= '0' && optarg[i] < '8') {
					config.output_limit[i] =
					    optarg[i] - '0';
				} else {
					errx(EXIT_FAILURE,
					     "main: output mask `%s' is illegal",
					     optarg);
				}
			}
			break;
		case OPT_WARN:
			strcpy(config.output_format, "a");
			config.warning = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_CRIT:
			strcpy(config.output_format, "a");
			config.critical = strtod_or_err(optarg, "illegal argument");
			break;
		case 'v':
			/* Print version */
			print_version();
		case 'h':
			/* Print help */
			usage(EXIT_SUCCESS);
		default:
			errx(EXIT_FAILURE,
			     "Try `%s --help' for more information.",
			     program_invocation_short_name);
		}
	}

	/* Output function selection */
	switch (config.output_format[0]) {
	case 't':
		output_analysis = output_txt;
		break;
	case 'a':
		output_analysis = output_alarming;
		break;
	case 'h':
		output_analysis = output_html;
		break;
	case 'H':
		output_analysis = output_html;
		fullhtml = true;
		break;
	case 'x':
		output_analysis = output_xml;
		break;
	case 'X':
		output_analysis = output_xml;
		break;
	case 'c':
		output_analysis = output_csv;
		break;
	default:
		errx(EXIT_FAILURE, "main: unknown output format `%c'",
		     config.output_format[0]);
	}

	/* Do the job */
	prepare_memory();
	parse_config(true, config.dhcpdconf_file, shared_networks);

	parse_leases();
	prepare_data();
	do_counting();
	tmp_ranges = xmalloc(sizeof(struct range_t) * num_ranges);
	if (sorts != 0) {
		mergesort_ranges(ranges, num_ranges, tmp_ranges);
	}
	if (config.reverse_order == true) {
		flip_ranges(ranges, tmp_ranges);
	}
	free(tmp_ranges);
	ret_val = output_analysis();

	clean_up();
	return (ret_val);
}

/* Global allocations, counter resets etc */
int prepare_memory(void)
{
	RANGES = 64;
	num_ranges = num_shared_networks = 0;
	shared_networks =
	    xmalloc(sizeof(struct shared_network_t) * SHARED_NETWORKS);

	ranges = xmalloc(sizeof(struct range_t) * RANGES);
	macaddr = NULL;

	/* First shared network entry is all networks */
	shared_networks->name = xstrdup("All networks");
	shared_networks->used = 0;
	shared_networks->touched = 0;
	shared_networks->backups = 0;

	return 0;
}
